File: src/lec17/Lec17Main.java
package lec17;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;

public class Lec17Main {
    public static void main(String[] args) {
        List<Fruit> fruits = Arrays.asList(
            new Fruit("사과", 1_000),
            new Fruit("사과", 1_200),
            new Fruit("사과", 1_500),
            new Fruit("바나나", 1_000),
            new Fruit("바나나", 1_500),
            new Fruit("바나나", 2_000),
            new Fruit("수박", 30_000)
        );

        Lec17Main main = new Lec17Main();
        List<Fruit> result = main.filterFruits(fruits, new FruitFilter() {
            @Override
            public boolean isSelected(Fruit fruit) {
                return Arrays.asList("사과", "바나나")
                    .contains(fruit.getName())
                    && fruit.getPrice() > 5000;
            }
        });

        List<Fruit> result2 = main.filterFruits2(fruits, fruit -> fruit.getName().equals("사과"));

        main.filterFruits3(fruits, Fruit::isApple);
    }

    private List<Fruit> findApplies(List<Fruit> fruits) {
        List<Fruit> applies = new ArrayList<>();
        for (Fruit fruit : fruits) {
            if (fruit.getName().equals("사과")) {
                applies.add(fruit);
            }
        }
        return applies;
    }

    private List<Fruit> findBananas(List<Fruit> fruits) {
        List<Fruit> bananas = new ArrayList<>();
        for (Fruit fruit : fruits) {
            if (fruit.getName().equals("바나나")) {
                bananas.add(fruit);
            }
        }
        return bananas;
    }

    private List<Fruit> findFruitsWithName(List<Fruit> fruits, String name) {
        List<Fruit> result = new ArrayList<>();

        for (Fruit fruit : fruits) {
            if (fruit.getName().equals(name)) {
                result.add(fruit);
            }
        }
        return result;
    }

    private List<Fruit> filterFruits(List<Fruit> fruits, FruitFilter fruitFilter) {
        List<Fruit> result = new ArrayList<>();
        for (Fruit fruit : fruits) {
            if (fruitFilter.isSelected(fruit)) {
                result.add(fruit);
            }
        }
        return result;
    }

    private List<Fruit> filterFruits2(List<Fruit> fruits, Predicate<Fruit> filter) {
        List<Fruit> result = new ArrayList<>();
        for (Fruit fruit : fruits) {
            if (filter.test(fruit)) {
                result.add(fruit);
            }
        }
        return result;
    }

    private List<Fruit> filterFruits3(List<Fruit> fruits, Predicate<Fruit> filter) {
        return fruits.stream()
            .filter(filter)
            .toList();
    }

}

File: src/lec17/Lec17Main.kt
package lec17

fun main() {
    val fruits = listOf(
        Fruit("사과", 1000),
        Fruit("사과", 1200),
        Fruit("사과", 1500),
        Fruit("바나나", 1000),
        Fruit("바나나", 1500),
        Fruit("바나나", 2000),
        Fruit("수박", 30000)
    )

    // 람다를 만드는 방법1
    val isApple1 = fun(fruit: Fruit): Boolean {
        return fruit.name == "사과"
    }

    // 람다를 만드는 방법1
    val isApple2: (Fruit) -> Boolean = { fruit: Fruit -> fruit.name == "사과"}


    // 람다를 직접 호출하는 방법1
    isApple1(fruits[0])

    // 람다를 직접 호출하는 방법2 - 명시적으로 invoke 사용
    isApple2.invoke(fruits[0])


    filterFruits(fruits, isApple1)

    // 소괄호 다음 중괄호 안에 바로 람다를 작성할 수 있다
    // 중괄호의 본문이 메서드의 가장 마지막 파라미터로 들어가게됨
    filterFruits(fruits) { fruit: Fruit -> fruit.name == "사과"}

    // 타입을 알고 있으므로 타입도 생략이 가능하며, 이름도 변경가능함
    filterFruits(fruits) { f -> f.name == "사과"}

    // 람다에서 사용하는 파라미터가 한 개인 경우 it 키워드 작성 가능
    filterFruits(fruits) { it.name == "사과"}

    filterFruits(fruits) { f ->
        println("출력")
        f.name == "사과"
    }

    var targetFruits = "바나나"
    targetFruits = "수박"
    filterFruits(fruits) {it.name == targetFruits}
}

private fun filterFruits(fruits: List<Fruit>, filter: (Fruit) -> Boolean): List<Fruit> {
    val result = mutableListOf<Fruit>()
    for (fruit in fruits) {
        if (filter(fruit)) {
            result.add(fruit)
        }
    }
    return result
}